API-Samples/Java/app/src/main/java/apisamples/TokensStore.java

```diff
@@ -0,0 +1,134 @@
+package apisamples;
+
+import apisamples.authentication.AccessTokenRequest;
+import apisamples.authentication.AccessTokenResponse;
+import apisamples.authentication.ArmAccessTokenPermission;
+import apisamples.authentication.ArmAccessTokenScope;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.identity.DefaultAzureCredential;
+import com.azure.identity.DefaultAzureCredentialBuilder;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpRequest;
+import java.text.MessageFormat;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import static apisamples.HttpUtils.Utils.httpStringResponse;
+import static apisamples.VideoIndexerClient.*;
+
+public class TokensStore {
+
+    private final LoadingCache<String, TimedEntry> cache;
+    private static final String armAccessTokenKey = "armAccessToken";
+    private static final String viAccountTokenKey = "viAccountToken";
+    private static final long ExpiryBufferTimeInMs = TimeUnit.MINUTES.toMillis(5);
+
+    private final Gson gson;
+    private final ArmAccessTokenPermission permission;
+    private final ArmAccessTokenScope scope;
+
+
+    public TokensStore(ArmAccessTokenPermission permission, ArmAccessTokenScope scope) {
+        this.cache = Caffeine.newBuilder()
+                    .build(key -> new TimedEntry(key, 0));
+        this.gson = new GsonBuilder().setPrettyPrinting().create();
+        this.permission = permission;
+        this.scope = scope;
+
+        //Generate the First ARM and VI Access Token and place them in cache
+        this.getVIAccessToken();
+    }
+
+
+    /**
+     * TimedEntry represents an expiry enabled cache item to be placed in the Loading Cache lib
+     */
+    private class TimedEntry {
+        long expiryTime;
+        String value;
+
+        public TimedEntry(String value, long expiryTimeEpocMilli) {
+            this.value = value;
+            this.expiryTime = expiryTimeEpocMilli;
+        }
+        public boolean isExpired() {
+            return System.currentTimeMillis() > expiryTime;
+        }
+    }
+
+    public String getArmAccessToken() {
+        return this.getOrCreate(armAccessTokenKey, this::generateArmAccessToken);
+    }
+
+    public String getVIAccessToken() {
+        return this.getOrCreate(viAccountTokenKey, this::generateVIAccessToken);
+    }
+
+    public String getOrCreate(String key, Supplier<TimedEntry> valueSupplier) {
+        TimedEntry timedEntry = cache.getIfPresent(key);

```

Comment:
I think you can get rid of all the logic by using the get

 /**
   * Returns the value associated with the {@code key} in this cache, obtaining that value from the
   * {@code mappingFunction} if necessary. This method provides a simple substitute for the
   * conventional "if cached, return; otherwise create, cache and return" pattern.
   * <p>
   * If the specified key is not already associated with a value, attempts to compute its value
   * using the given mapping function and enters it into this cache unless {@code null}. The entire
   * method invocation is performed atomically, so the function is applied at most once per key.
   * Some attempted update operations on this cache by other threads may be blocked while the
   * computation is in progress, so the computation should be short and simple, and must not attempt
   * to update any other mappings of this cache.
   * <p>
   * <b>Warning:</b> as with {@link CacheLoader#load}, {@code mappingFunction} <b>must not</b>
   * attempt to update any other mappings of this cache.
   *
   * @param key the key with which the specified value is to be associated
   * @param mappingFunction the function to compute a value
   * @return the current (existing or computed) value associated with the specified key, or null if
   *         the computed value is null
   * @throws NullPointerException if the specified key or mappingFunction is null
   * @throws IllegalStateException if the computation detectably attempts a recursive update to this
   *         cache that would otherwise never complete
   * @throws RuntimeException or Error if the mappingFunction does so, in which case the mapping is
   *         left unestablished
   */
  @PolyNull
  V get(K key, Function<? super K, ? extends @PolyNull V> mappingFunction);

Reply:
I took another look, I'm not sure it's possilbe since Caffeine does not have eviction policy per record :(

Reply:
where is this code coming from ?